### PR TITLE
Protocol Documentation

### PR DESCRIPTION
**Pull request by DavidVorick**
The protocol is nearing a final state, and the consensus documentation now reflects that. I tried to make Consensus.md comprehensive (and proper markdown), but there are 4 things left to document:

1. The specification for encoding structs into strings (that get hashed)
2. The specification for deriving block id + merkle root, miner output id, siacoin output id, contract id, storage proof output id, siafund output id, siafund claim output id.
3. Specify which crypto is in use (hash algo, signature algo)
4. Pick a timestamp for the genesis block.

Everything else should be specified, although not perfectly followed by the codebase yet. The codebase does not follow rules for:

- siacoins having 128bits
- siafunds are not followed at all
- outputs  that can't be spent for 100 blocks
- limitations on storage proof transactions

Other than that, we're pretty much done with the protocol, and it's fairly polished.

### CODE REVIEW

**Comment by DavidVorick** [comment]
uh. For some reason we lost the other PR again? Maybe it's because I'm pushing to the same branch that the other PR was on and that's confusing Github? =/

File: Consensus.md [hunk 1]
```diff
@@ -1,48 +1,324 @@
...
+blocks, which can be compared to the 4x clamp of Bitcoin. The amount of work
+required to quadruple the difficulty in Sia is 3000x the starting difficulty,
+which can be compared to 2016x for Bitcoin. The amount of work required to 16x
+the difficulty in Sia is 15,000x the original difficulty, which can be compared
+to 10,000x for Bitcoin.
+
+Block Subsidy
+-------------
+
+The coinbase for a block is (300,000 - (1 * height)) * 2^80, with a minimum of
```
Comment (lukechampine) [inline Consensus.md]: 1 * height??

File: Consensus.md [hunk 2]
```diff
@@ -1,48 +1,324 @@
...
+- Outputs
+- File Contracts
+- Storage Proofs
+- Siafund Inputs
+- Siafund Outputs
+- Arbitrary Data
+- Signatures
+
+The financial sum of all the inputs must equal the financial sum of all the
+miner fees, outputs, and contract payouts. There can be no leftovers.
```
Comment (lukechampine) [inline Consensus.md]: "financial" is probably redundant here. At most I would use "monetary," and only for the first instance (the rest being implied).

File: Consensus.md (hunk 1 above)
Comment (DavidVorick) [inline Consensus.md]: :P

File: Consensus.md [hunk 3]
```diff
@@ -1,48 +1,324 @@
...
+Sia is a cryptosystem that uses a Bitcoin style Proof of Work blockchain to
+achieve network consensus. The rules of consensus are detailed in this
+document. The most detailed and accurate explanation however is the codebase
+itself. If you are looking for a more precise understanding of the consensus
+rules, a good place to start is consensus/types.go, and another good starting
+place is the function AcceptBlock, which can be found in consensus/blocks.go.
+
+This document will be more understandable if you have a general understanding
+of proof of work blockchains, and does not try to build up from first
+priciples.
```
Comment (lukechampine) [inline Consensus.md]: principles

File: Consensus.md [hunk 4]
```diff
@@ -1,48 +1,324 @@
...
+File Contracts
+--------------
+
+A file contract is an agreement by some party to prove they have a file at a
+given point in time. The contract contains the merkle root of the data being
+stored, and the size in bytes of the data being stored.
+
+The merkle root is formed by breaking the file into 64 byte segments and
+hashing each segment to form the leaves of the merkle tree. The final segment
+is not padded out.
```
Comment (lukechampine) [inline Consensus.md]: Merkle is always capitalized

File: Consensus.md [hunk 5]
```diff
@@ -1,48 +1,324 @@
...
+
+When creating a storage proof, you only prove that you have a single 64 byte
+segment of the file. The piece that you must prove you have is chosen
+psuedorandomly using the contract id and the id of the 'trigger block'.  The
+trigger block is the block at height 'Start' - 1, where 'Start' is the value
+'Start' in the contract that the storage proof is fulfilling.
+
+The file is composed of 64 byte segments whose hashes compose the leaves of a
+merkle tree. When proving you have the file, you must prove you have one of the
+leaves. To determine which leave, take the hash of the contract id concatenated
```
Comment (lukechampine) [inline Consensus.md]: leaf

File: Consensus.md [hunk 6]
```diff
@@ -1,48 +1,324 @@
...
+output. These spend conditions must be signed.
+
+A special output is created when a siafund is used as input. All of the
+siacoins that have accrued in the siafund since its last spend are sent to the
+'Claim Destination' found in the siafund output, which is a normal siacoin
+address. The number of accrued siacoins is counted by taking the size of the
+siacoin pool when the output was created and comparing it to the current size
+of the siacoin pool. The equation is:
+
+	(Current Pool Size - Previous Pool Size) / 10,000 * siafund quantity
```
Comment (lukechampine) [inline Consensus.md]: need one more set of parens here, for clarity

File: Consensus.md [hunk 7]
```diff
@@ -1,48 +1,324 @@
...
+
+Genesis Set
+-----------
+
+The genesis block will be a block with a timestamp of TODO:TBD. All other
+fields will be empty. The required target for the next block shall be [0, 0, 0,
+8, 0...], where each value is a byte.
+
+The genesis block does not need to meet a particular target, and indeed it does
+not.
```
Comment (lukechampine) [inline Consensus.md]: "indeed it does not" is meaningless, no?

Comment (DavidVorick) [inline Consensus.md]: lol. It meets the [255, 255...] target, at the very least.

**Pull request merged**